File: lib/nb2mdx.py
import base64
from pathlib import Path

import nbformat


def wrap_html(html_str):
    # Escape backticks and ensure proper JS string formatting
    html_str = html_str.replace("`", "\\`")
    return f'<div className="nboutput-container" dangerouslySetInnerHTML={{ {{ __html: `{html_str}` }} }} />'


def save_image(image_data, filename: str, path: Path):
    # Save the image data to a file
    path.mkdir(parents=False, exist_ok=True)
    with open(path / filename, "wb") as f:
        # Decode base64 string and write to file
        f.write(base64.b64decode(image_data))


def extract_outputs(outputs, output_parent: Path, nb_name: str):
    md_parts = []
    figure_path = output_parent.joinpath("nbfigures")

    for output in outputs:
        if output.output_type == "stream":
            md_parts.append(f"\n```text\n{output.text}\n```")
        elif output.output_type in {"execute_result", "display_data"}:
            data = output.get("data", {})
            if "text/html" in data:
                md_parts.append(wrap_html(data["text/html"]))
                continue
            if "image/png" in data:
                b64 = data["image/png"]
                filename = f"{nb_name}_img{len(md_parts)}.png"
                save_image(b64, filename, figure_path)
                md_parts.append(f"![{filename}](./nbfigures/{filename})")
                continue
            if "text/plain" in data:
                md_parts.append(f"\n```text\n{data['text/plain']}\n```")
    return "\n".join(md_parts)


def convert_ipynb_to_mdx(input_path, output_path):
    nb = nbformat.read(input_path, as_version=4)
    output_lines = []
    output_parent = Path(output_path).parent
    output_path = Path(output_path)

    for cell in nb.cells:
        if cell.cell_type == "raw":
            output_lines.append(cell.source)
        elif cell.cell_type == "markdown":
            output_lines.append(cell.source)
        elif cell.cell_type == "code":
            # check if first line is a comment
            title = ""
            if cell.source.startswith("#"):
                title, cell.source = cell.source.split("\n", 1)
                title = title.strip("#").strip()
            output_lines.append(
                f"```python{' title="{}"'.format(title) if title else ''}\n{cell.source}\n```"
            )
            if cell.outputs:
                output_lines.extend(
                    [
                        '<div className="nboutput">\n',
                        extract_outputs(cell.outputs, output_parent, output_path.stem),
                        "\n</div>",
                    ]
                )

    Path(output_path).write_text("\n\n".join(output_lines), encoding="utf-8")
    print(f"Converted: {input_path} → {output_path}")


if __name__ == "__main__":
    import sys

    if len(sys.argv) != 3:
        print("Usage: python nb_to_mdx.py input.ipynb output.mdx")
    else:
        convert_ipynb_to_mdx(sys.argv[1], sys.argv[2])

File: nb2mdx/nb2mdx.ts
import fs from "fs";
import crypto from "crypto";
import path from "path";
import AnsiToHtml from "ansi-to-html";

function wrapHtml(html: string[]) {
    const escaped = html.join("").replace(/`/g, "\\`");
    return `<div className="nboutput-html" dangerouslySetInnerHTML={{ __html: \`${escaped}\` }} />`;
}

const ansi = new AnsiToHtml();

function renderStreamAsHtml(textLines: string[]) {
    const raw = textLines.join("");
    const html = ansi.toHtml(raw);
    return `<div className="nboutput-terminal" dangerouslySetInnerHTML={{ __html: \`${html.replace(/`/g, "\\`")}\` }} />`;
}

const FIGURE_DIR = path.resolve("content/docs/.nbfigures");
fs.mkdirSync(FIGURE_DIR, { recursive: true });

function saveImageBase64(base64: any, altText: string): string {
    const hash = crypto.createHash("sha1").update(base64).digest("hex");
    const filename = `${hash}.png`;
    const filePath = path.join(FIGURE_DIR, filename);
    const altTextClean =
        altText.replace("<", "&lt;").replace(">", "&gt;") ?? "output";

    if (!fs.existsSync(filePath)) {
        fs.writeFileSync(filePath, Buffer.from(base64, "base64"));
    }

    return `\n![${altTextClean}](content/docs/.nbfigures/` + filename + ")\n";
}

function extractOutputs(outputs: any[]) {
    const parts = [];

    for (const output of outputs) {
        if (output.output_type === "stream") {
            parts.push(`\n\`\`\`\n${output.text}\n\`\`\``);
        } else if (
            ["display_data", "execute_result"].includes(output.output_type)
        ) {
            const data = output.data || {};
            if (data["text/html"]) {
                parts.push(wrapHtml(data["text/html"]));
            } else if (data["image/png"]) {
                const img_string = saveImageBase64(
                    data["image/png"],
                    data["text/plain"][0] ?? "",
                );
                parts.push(img_string);
            } else if (data["text/plain"]) {
                parts.push(`\n\`\`\`\n${data["text/plain"]}\n\`\`\``);
            }
        }
    }

    return parts.join("\n\n");
}

function convertNotebook(inputPath: string, outputPath: string) {
    const nb = JSON.parse(fs.readFileSync(inputPath, "utf-8"));
    const lines = [];

    for (const cell of nb.cells) {
        if (["markdown", "raw"].includes(cell.cell_type)) {
            lines.push(cell.source.join(""));
        } else if (cell.cell_type === "code") {
            // if first line is a comment, parse it as the title
            const firstLine = cell.source[0].trim();
            let title = null;

            if (firstLine.startsWith("#")) {
                title = firstLine.replace(/^#+/, "").trim();
                cell.source.shift(); // remove the title line
            }
            lines.push(
                "```python " +
                (title ? `title="${title}"` : "") +
                "\n" +
                cell.source.join("") +
                "\n```",
            );
            if (cell.outputs?.length) {
                lines.push(
                    [
                        '<div className="nboutput">',
                        extractOutputs(cell.outputs),
                        "</div>",
                    ].join("\n"),
                );
            }
        }
    }

    fs.writeFileSync(outputPath, lines.join("\n\n"), "utf-8");
    console.log(`✅ Converted: ${outputPath}`);
}

// CLI
if (process.argv.length !== 4) {
    console.error("Usage: node convertNotebookToMdx.js input.ipynb output.mdx");
    process.exit(1);
}

const [input, output] = process.argv.slice(2);
convertNotebook(input, output);
